staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache.go

```diff
@@ -0,0 +1,135 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	resourceapi "k8s.io/api/resource/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	resourceinformers "k8s.io/client-go/informers/resource/v1"
+	klog "k8s.io/klog/v2"
+)
+
+// ExtendedResourceCache maintains a global cache of extended resource to device class mappings
+// that can also be accessed by event handlers without requiring an extra scheduling cycle.
+type ExtendedResourceCache struct {
+	mutex sync.RWMutex
+	// mapping maps extended resource name to device class name
+	mapping map[v1.ResourceName]string
+	logger  klog.Logger
+}
+
+// NewExtendedResourceCache creates a new ExtendedResourceCache instance and populates it
+// with the current device classes from the informer cache.
+func NewExtendedResourceCache(deviceClassInformer resourceinformers.DeviceClassInformer, logger klog.Logger) *ExtendedResourceCache {
+	cache := &ExtendedResourceCache{
+		mapping: make(map[v1.ResourceName]string),
+		logger:  logger,
+	}
+
+	// Populate the cache with existing device classes from the informer
+	if deviceClassInformer != nil {
+		deviceClassLister := deviceClassInformer.Lister()
+		classes, err := deviceClassLister.List(labels.Everything())
+		if err != nil {
+			logger.Error(err, "Failed to populate extended resource cache during initialization")
+		} else {
+			for _, deviceClass := range classes {
+				cache.updateMapping(deviceClass)
+			}
+			logger.V(4).Info("Created and populated extended resource cache", "mappingCount", len(cache.mapping))
+		}
+	} else {
+		logger.V(4).Info("Created extended resource cache without initial population (deviceClassInformer is nil)")
+	}
+
+	return cache
+}
+
+// GetDeviceClass returns the device class name for the given extended resource name.
+// Returns empty string if the resource name is not found in the cache.
+func (c *ExtendedResourceCache) GetDeviceClass(resourceName v1.ResourceName) string {
+	if c == nil {
+		return ""
+	}
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	return c.mapping[resourceName]
+}
+
+// OnAdd handles the addition of a new device class.
+func (c *ExtendedResourceCache) OnAdd(obj interface{}, isInInitialList bool) {
+	if deviceClass, ok := obj.(*resourceapi.DeviceClass); ok {
+		c.updateMapping(deviceClass)
+	}
+}
+
+// OnUpdate handles updates to an existing device class.
+func (c *ExtendedResourceCache) OnUpdate(oldObj, newObj interface{}) {
+	if deviceClass, ok := newObj.(*resourceapi.DeviceClass); ok {
+		c.updateMapping(deviceClass)
+	}
+}
+
+// OnDelete handles deletion of a device class.
+func (c *ExtendedResourceCache) OnDelete(obj interface{}) {
+	if deviceClass, ok := obj.(*resourceapi.DeviceClass); ok {
```

Comment:
informer may send DeletedFinalStateUnknown instead of the deviceClass, so this case also needs to be handled, please check out how other controllers handle it. It should look something like below:

	deviceClass, ok := obj.(*resourceapi.DeviceClass)
	if !ok {
		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
		if !ok {
			utilruntime.HandleError(fmt.Errorf("couldn't get object from tombstone %#v", obj))
			return
		}
		deviceClass, ok = tombstone.Obj.(*resourceapi.DeviceClass)
		if !ok {
			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a DeviceClass %#v", obj))
			return
		}
	}